### ISSUE TITLE
Support for no-frills FP8 matmuls

### ISSUE THREAD

**Issue by snarayan21** [pytorch/pytorch#123761]
### 🚀 The feature, motivation and pitch

Doing matmuls in FP8 is essential for getting the best performance out of the newest hardware (e.g. H100s and beyond). Currently, the only option for doing matmuls in FP8 is through `_scaled_mm`, which includes overhead like calculating absolute maximums, taking in a bunch of scale factors, etc, slowing it down and eroding the benefits of using FP8. What would be amazing is if there was full, accessible support for FP8 matmuls -- specifically through `torch.addbmm()` or more generally through `torch.matmul()`, where:
* The accumulation/output dtype can be specified or controlled somehow
* Potentially, an output scale factor may be applied

This would let users harness the full efficiency of FP8 math, and choosing/handling scaling factors and such could be left to the user as well.

**Comment by awgu** [comment]
cc: @vkuzo @drisspg

**Comment by albanD** [comment]
Hey!

Have you used the proposal from https://github.com/pytorch-labs/float8_experimental ?
Or you're looking for lower level APIs that you would control more on your end?

**Comment by snarayan21** [comment]
Hey @albanD, I've read through and used the float8 experimental repo pretty extensively and as you say, we'd really like something lower level that we can control. The main win from fp8 is speed, and without that lower-level customizability, getting closer to the ideal speedup is not possible.

The float8 experimental library shows speedups of ~1.2x which we have seen internally as well, but with TE we have seen 1.5-1.6x. Lower-level APIs directly through PyTorch would allow us to get even higher, and let us customize in accordance with our use cases.

**Comment by drisspg** [comment]
Hey @snarayan21 would you be able to provide a reference or some code pointer for the speed up numbers you mentioned? 
From the initial post its seems like you want a scaled_mm but no input scales only an output scale? This sounds like addmm but without an input parameter? Also out of curiosity is there anything foundational that is preventing you from creating this custom op and calling into it?

**Comment by snarayan21** [comment]
Hey @drisspg I'm getting the float8_experimental numbers from the repo readme [here](https://github.com/pytorch-labs/float8_experimental/tree/main) and the TE numbers are from our recently released blogpost [here](https://www.databricks.com/blog/turbocharged-training-optimizing-databricks-mosaic-ai-stack-fp8) -- correction, we're seeing 1.4-1.5x with TE, apologies.

Nothing is really preventing me from making a custom op, but I think it would be a great feature to have first-class support for this in pytorch itself, both for us and for other users. Additionally, having the op work seamlessly with torch compile and other torch features would be great too.

**Comment by vkuzo** [comment]
> and the TE numbers are from our recently released blogpost [here](https://www.databricks.com/blog/turbocharged-training-optimizing-databricks-mosaic-ai-stack-fp8) -- correction, we're seeing 1.4-1.5x with TE, apologies.

Just curious, is the 1.4x to 1.5x just from float8, or float8 + other optimizations?  I see two mentions of this in the blog and just wanted to ensure I'm understanding correctly.
* "We achieve 1.4x-1.5x speedup with FP8 compared to BF16."
* "With a combination of FP8 and other optimizations, we are able to achieve ~1.5x better training throughput across different model sizes."

**Comment by vkuzo** [comment]
> Currently, the only option for doing matmuls in FP8 is through _scaled_mm, which includes overhead like calculating absolute maximums, taking in a bunch of scale factors, etc, slowing it down and eroding the benefits of using FP8. What would be amazing is if there was full, accessible support for FP8 matmuls

I agree with this and we are also thinking about how to make the float8 matmul support more flexible.  Today, `torch._scaled_mm` is hooked up to the cuBLAS float8 enabled matmul.  From measurements we have done, the `torch._scaled_mm` op itself is consistently fast and people often see a ~2x compared to the bf16 matmul.  We don't have evidence at the moment that calculation of the output amax is slow.  The slowdowns we have observed have been from calculating the scaling factors (currently only per-tensor is supported) and the extra memory writes/reads to cast the inputs to float8 since that isn't fused into the matmul at the moment.

Some things we are thinking about:
0. adding a cutlass or triton backend to the float8 matmul, to enable more flexibility such as the additional features below
1. enabling scaling per-axis, in addition to the currently supported per-tensor
2. improving the fusion for the cast of inputs to float8, whether to the preceding ops or into the matmul

**Comment by snarayan21** [comment]
> Just curious, is the 1.4x to 1.5x just from float8, or float8 + other optimizations? I see two mentions of this in the blog and just wanted to ensure I'm understanding correctly.
@vkuzo The speedup is with other optimizations as well, but the bulk of the speedup is from using fp8.

From my own profiling on a real transformer model (around 1B params), it seems that using `torch._scaled_mm` is around 5-15% slower than TE's `te_gemm` [function](https://github.com/NVIDIA/TransformerEngine/blob/5d34b2ac9d88a853b864e4c3cf7098732d0ed601/transformer_engine/pytorch/csrc/extensions/gemm.cu#L10), but there are some gotchas and incompatibilities with torch compile when combining `te_gemm` with `float8_experimental.Float8Tensor`. The fact that `_scaled_mm` and `Float8Tensor` are compatible with torch compile is great but it seems that the performance isn't quite optimal. The `te_gemm` also seems to support matrices of different layouts, while `_scaled_mm` only supports `A@B` for row-major `A` and col-major `B`, at least according to [this comment](https://github.com/pytorch/pytorch/blob/8ce29f1416f6a3852dc10bfe9326afc6cca4c2f0/aten/src/ATen/native/cuda/Blas.cpp#L799). Not sure if this will impact performance too much.

> the extra memory writes/reads to cast the inputs to float8 since that isn't fused into the matmul at the moment.
I've seen the same thing when profiling as well -- having some fused ops to combine casts with matmuls, or casts with transposes, would be great. For example TE has a `fused_cast_transpose` op [here](https://github.com/NVIDIA/TransformerEngine/blob/5d34b2ac9d88a853b864e4c3cf7098732d0ed601/transformer_engine/pytorch/csrc/extensions/transpose.cu#L9), among others, that seem to make the FP8 linear layers faster.

> adding a cutlass or triton backend to the float8 matmul
This would be great. Would also make it more accessible for me and others to build custom functionality :)

We'd love to be involved with the development of new fp8 features, and would be excited to test them out too!

**Comment by vkuzo** [comment]
> it seems that using torch._scaled_mm is around 5-15% slower than TE's te_gemm [function](https://github.com/NVIDIA/TransformerEngine/blob/5d34b2ac9d88a853b864e4c3cf7098732d0ed601/transformer_engine/pytorch/csrc/extensions/gemm.cu#L10)

thanks for pointing this out!  This is unexpected to me and we should understand why that is, as both TE and float8_experimental seem to call into cuBLAS for the actual gemm.  There might be a bug in how we are setting one of the arguments from the PyTorch side.  cc @malfet , @drisspg

**Comment by snarayan21** [comment]
This was just one one model size, so I will benchmark more mm sizes and update with more info.

**Comment by snarayan21** [comment]
@vkuzo Benchmarked Float8Tensor fp8 square matmuls with `_scaled_mm` (called through `torch.matmul()`) vs the `te_gemm` (through a simple wrapper) -- results are below. At small matrix sizes, the two functions are comparable, but it seems that as matrices get bigger, `_scaled_mm` becomes worse than `te_gemm`, and we only really care about relatively large matmuls. Both methods do not get to the peak H100 performance of 2000 TFLOPs.

```
Scaled MM vs TE GEMM benchmarking, 50 trials each

Dim	Scaled MM Mean 	Scaled MM Std	TE GEMM Mean 	TE GEMM Std
---------------------------------------------------------------------------------
16	1.6990ms	11.2934ms	0.2527ms	1.1087
32	0.0909ms	0.0409ms	0.1009ms	0.0375
64	0.0878ms	0.0108ms	0.0954ms	0.0071
128	0.0975ms	0.0253ms	0.1006ms	0.0116
256	0.1201ms	0.0128ms	0.1161ms	0.0092
512	0.1581ms	0.0502ms	0.1272ms	0.0145
1024	0.1498ms	0.0107ms	0.1260ms	0.0106
2048	0.2223ms	0.0135ms	0.1655ms	0.0296
4096	0.3973ms	0.0114ms	0.2951ms	0.0301
8192	1.0274ms	0.0295ms	0.9037ms	0.0705
16384	6.5939ms	0.0143ms	5.9866ms	0.0985
```

**Comment by snarayan21** [comment]
@drisspg @vkuzo Implementing customized fp8 matmuls in triton is possible, but
1) there doesn't seem to be full support from torch.compile
2) the latest pytorch nightly uses a triton version that seems to be older than the latest triton nightly.

I am going to file a separate issue for 1). For 2), the more recent triton nightlies give a pretty nice boost to the FP8 matmul performance for non-square matrix sizes (~50% faster for small matrices, ~10-20% for larger ones). How often does PyTorch update its triton version in the `pytorch-triton` wheels, and when could I anticipate a newer version of `pytorch-triton`?

Fusing casts into the matmul op is also running into some difficulties -- see [this issue](https://github.com/openai/triton/issues/2781) on the triton side.

**Comment by snarayan21** [comment]
@drisspg thanks for the pointer -- that does seem to speed up the benchmarking! I didn't do that earlier because when I enabled `use_fast_accum` and benchmarked actual model training, it was actually slower than disabling fast accum. Which is still puzzling, but the benchmark numbers look good. For some reason, the TE Gemm is slower then using `use_fast_accum`...

```
Dim  Scaled MM, ms  TE GEMM, ms  Triton MM, ms  Scaled MM, TFLOPS  TE GEMM, TFLOPS  Triton MM, TFLOPS
-----------------------------------------------------------------------------------------------------------------------------------
256	0.0172ms	0.0203ms	0.0075ms	1.9564		1.6558		4.4465
512	0.0184ms	0.0207ms	0.0081ms	14.6263		12.9672		33.3241
1024	0.0210ms	0.0222ms	0.0105ms	102.1206	96.5827		204.2094
2048	0.0292ms	0.0351ms	0.0198ms	587.9880	489.8969	865.7288
4096	0.1103ms	0.1278ms	0.1102ms	1246.3625	1075.5971	1246.6477
6144	0.3485ms	0.4041ms	0.3632ms	1330.8981	1147.9393	1277.1748
8192	0.8401ms	0.9381ms	0.8636ms	1308.7337	1172.0446	1273.2081
12288	2.8396ms	3.1324ms	2.8604ms	1306.8055	1184.6544	1297.3107
16384	6.8239ms	7.2710ms	6.8430ms	1289.0155	1209.7492	1285.4120
```

**Comment by Xynonners** [comment]
sorry to jump on this thread, but I'd really appreciate float8 being hooked up to bmm as well.
